Multithreaded LSP Wrapper

jvilk-stripe opened the pull request:
<!-- (optional) Explain your change, focusing on the details of the solution. This is a great place to call out user-visible changes. -->

Adds a multithreaded mode to LSP wrapper.

Overview:

* **Introduces subclasses for LSPWrapper for single-threaded and multi-threaded operations..** 
* **Single threaded interface is unchanged.** You can give LSPWrapper a set of messages, and it will return all messages created in response to that message.
* **Multi-threaded interface decouples sending from receiving.** You `send` messages to LSP, and blockingly `read` new ones.
* **Both LSPWrapper classes now contain factory functions rather than multiple constructors.** This change let me mark many fields on LSPWrapper as `const`. The constructors are now private.
* **Introduces a _sorbet/fence_ instruction that lets multithreaded tests clear the pipeline.** Makes it possible to isolate responses to a specific message by inserting fences.
* **Adds a first multithreaded LSPWrapper test.**

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

Required to easily test the preemptive slow path end-to-end.

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated test.

jvilk-stripe commented:
This is now ready to review!

jvilk-stripe commented on main/lsp/wrapper.h:
```diff
@@ -3,32 +3,45 @@
 class LSPConfiguration;
+
 class LSPWrapper final {
-    static const std::string EMPTY_STRING;
```
No longer needed; the default constructor for `string_view` constructs a zero-length string view.

jvilk-stripe commented on test/lsp/BUILD:
```diff
@@ -30,6 +30,30 @@ cc_test(
 )
 
+cc_test(
+    name = "multithreaded_protocol_test_corpus",
```
I wonder if there's a cleaner way to define these. I'm basically using the same build logic for all of these .cc files.

jvilk-stripe commented on test/lsp/ProtocolTest.cc:
```diff
@@ -10,8 +10,17 @@ void ProtocolTest::SetUp() {
+    auto opts = make_shared<realmain::options::Options>();
+    opts->disableWatchman = true;
+    if (useMultithreading) {
+        // TODO: Add async ability to ProtocolTest. For now, send methods on wrapper aren't available.
```
I'll need to add this in for preemptive slow path changes.

jvilk-stripe commented on test/lsp/protocol_test_corpus.cc:
```diff
@@ -10,7 +10,7 @@ using namespace std;
 
 // Adds two new files that have errors, and asserts that Sorbet returns errors for both of them.
-TEST_F(ProtocolTest, AddFile) {
+TEST_P(ProtocolTest, AddFile) {
```
TEST_P is required for parameterized tests 🤷‍♂

jvilk-stripe commented on main/lsp/wrapper.h:
```diff
@@ -40,32 +53,89 @@ class LSPWrapper final {
+     * Raw constructor; use create* factory methods to construct the wrapper. Note: Constructor is unwieldy so we can
+     * make class fields `const`.
+     */
+    LSPWrapper(std::shared_ptr<options::Options> opts, std::shared_ptr<LSPLoop> lspLoop,
```
I wanted to make this private, but then that disallows `make_unique`. Is there an alternative here? (Maybe I have to use `new` and move it into a `unique_ptr`?)

DarkDimius replied:
> Maybe I have to use new and move it into a unique_ptr

yes, this is the alternative

aisamanra replied:
It looks like the abseil recommendation is a combination of `new` and `absl::WrapUnique` (c.f. [here](https://abseil.io/tips/134)) but this is outside the area of C++ where I trust my own knowledge; it might be worth asking others on the team if they have a preference.

DarkDimius commented on main/lsp/wrapper.cc:
```diff
@@ -1,75 +1,254 @@
 
 namespace sorbet::realmain::lsp {
+namespace {
+void fixGS(core::GlobalState &gs, options::Options &opts) {
```
what does this do and why?
Can we give it more expressive name?

DarkDimius commented on main/lsp/wrapper.h:
```diff
@@ -40,32 +53,89 @@ class LSPWrapper final {
     std::vector<std::unique_ptr<LSPMessage>> getLSPResponsesFor(std::vector<std::unique_ptr<LSPMessage>> messages);
 
+    /**
+     * [Multithreaded only] Sends one message to LSP. Responses are passed asynchronously to the `processResponse`
```
rather than writing it in docs, can we split this class into multiple classes that have a common interface?

jvilk-stripe replied:
Sure; so both implement the shared interface, and only the multithreaded variant contains the `send` methods? That seems like a good idea to me.

DarkDimius commented on main/lsp/wrapper.h:
```diff
@@ -40,32 +53,89 @@ class LSPWrapper final {
 
     /**
      * Send a message to LSP, and returns any responses.
+     * NOTE: In multithreaded mode, the responses to these messages are _not_ sent to `processResponse`!
```
rather than having a method that does things differently, can we have two methods with different names?

jvilk-stripe replied:
? `send` and `getLSPResponsesFor` do have different names?

jvilk-stripe replied:
Talked IRL; going to proceed with a two-class setup where:

* There is no multithreaded implementation of `getLSPResponsesFor`. Tests will use a test function that emulate this functionality.
* We don't pass a lambda to LSPWrapper. Instead, messages will queue internally and LSPWrapper will expose a blocking `read` method.

DarkDimius commented on main/lsp/wrapper.h:
```diff
@@ -40,32 +53,89 @@ class LSPWrapper final {
+
     // N.B.: Sorbet assumes we 'own' this object; keep it alive to avoid memory errors.
-    options::Options opts;
+    const std::shared_ptr<options::Options> opts;
```
+1 for making these fields const

DarkDimius commented on main/lsp/wrapper.cc:
```diff
@@ -1,75 +1,254 @@
+    opts.runLSP = true;
+}
+
+enum class FenceState {
```
I didn't carefully check that graph of state transitions and how it works with concurrency/implications of it.

DarkDimius commented:
https://github.com/sorbet/sorbet/pull/2200#discussion_r348722470
https://github.com/sorbet/sorbet/pull/2200#discussion_r348722632
are the ones I'd prefer addressed, all the others are minor

aisamanra commented on main/lsp/wrapper.h:
```diff
@@ -3,32 +3,45 @@
+    static std::unique_ptr<LSPWrapper>
+    createInternal(std::string_view rootPath, std::shared_ptr<options::Options> options,
+                   std::optional<std::function<void(std::unique_ptr<LSPMessage>)>> &&processResponse,
+                   int numWorkerThreads, bool disableFastPath);
```
It's not a big deal in any way, but is there a reason these have to have the same name? Given their arguments, they don't seem like they're meaningfully overloaded versions of each other; it might be nice to be able to clearly disambiguate them by name instead of just by argument structure.

jvilk-stripe replied:
No longer a problem in the recent changes.

jvilk-stripe commented:
I believe I've addressed all feedback.

There are now subclasses of `LSPWrapper`: One is single-threaded, the other is multithreaded. They have different APIs to reflect that reality.

The multithreaded wrapper no longer accepts a lambda. Instead, it exposes a blocking `read` method. As a bonus, this lets me use the same `LSPOutput` class for both versions of `LSPWrapper`, and avoids the state machine I had earlier.

jvilk-stripe commented on test/helpers/lsp.cc:
```diff
@@ -347,4 +347,54 @@ unique_ptr<LSPMessage> makeChange(std::string_view uri, std::string_view content
     return make_unique<LSPMessage>(move(didChangeNotif));
 }
 
+vector<unique_ptr<LSPMessage>> getLSPResponsesFor(LSPWrapper &wrapper, vector<unique_ptr<LSPMessage>> messages) {
```
Tests use this method so we can share logic that operates in both single and multithreaded mode.

jvilk-stripe commented:
I'm fixing the build error right now; ~WorkerPool assumes that logger is still alive, but it gets destructed with LSPConfiguration.

jvilk-stripe commented:
Rebased on master; this should be ready to review now!
